tracy-client/src/gpu.rs

```diff
@@ -0,0 +1,163 @@
+use std::sync::atomic::{AtomicU16, AtomicU8, Ordering};
+
+use crate::Client;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+struct ___tracy_gpu_new_context_data {
+    gpu_time: i64,
+    period: f32,
+    context: u8,
+    flags: u8,
+    type_: u8,
+}
+
+#[repr(u8)]
+///
+pub enum GpuContextType {
+    ///
+    Invalid,
+    ///
+    OpenGL,
+    ///
+    Vulkan,
+    ///
+    OpenCL,
+    ///
+    Direct3D12,
+    ///
+    Direct3D11,
+}
+
+///
+pub struct GpuContext {
+    #[cfg(feature = "enable")]
+    _client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+}
+#[cfg(feature = "enable")]
+static GPU_CONTEXT_INDEX: AtomicU8 = AtomicU8::new(0);
+
+///
+pub struct GpuSpan {
+    #[cfg(feature = "enable")]
+    _client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+    #[cfg(feature = "enable")]
+    start_query_id: u16,
+    #[cfg(feature = "enable")]
+    end_query_id: Option<u16>,
+}
+#[cfg(feature = "enable")]
+static GPU_SPAN_INDEX: AtomicU16 = AtomicU16::new(0);
+
+impl Client {
+    ///
+    pub fn new_gpu_context(
+        self,
+        name: Option<&str>,
+        ty: GpuContextType,
+        gpu_timestamp: i64,
+        period: f32,
+    ) -> GpuContext {
+        #[cfg(feature = "enable")]
+        unsafe {
+            let context = GPU_CONTEXT_INDEX.fetch_add(1, Ordering::Relaxed);
+
+            sys::___tracy_emit_gpu_new_context(std::mem::transmute(
+                ___tracy_gpu_new_context_data {
+                    gpu_time: gpu_timestamp,
+                    period,
+                    context,
+                    flags: 0,
+                    type_: ty as u8,
+                },
+            ));
+
+            if let Some(name) = name {
+                sys::___tracy_emit_gpu_context_name_serial(sys::___tracy_gpu_context_name_data {
+                    context,
+                    name: name.as_ptr() as *const i8,
+                    len: name.len().min(u16::MAX as usize) as u16,
+                })
+            }
+
+            GpuContext {
+                _client: self,
+                context,
+            }
+        }
+        #[cfg(not(feature = "enable"))]
+        GpuContext {}
+    }
+}
+
+impl GpuContext {
+    ///
+    pub fn span_alloc(&self, name: &str, function: &str, file: &str, line: u32) -> GpuSpan {
+        #[cfg(feature = "enable")]
+        unsafe {
+            let srcloc = sys::___tracy_alloc_srcloc_name(
+                line,
+                file.as_ptr().cast(),
+                file.len(),
+                function.as_ptr().cast(),
+                function.len(),
+                name.as_ptr().cast(),
+                name.len(),
+            );
+
+            let query_id = GPU_SPAN_INDEX.fetch_add(1, Ordering::Relaxed);
+
+            sys::___tracy_emit_gpu_zone_begin_alloc_serial(sys::___tracy_gpu_zone_begin_data {
+                srcloc,
+                queryId: query_id,
+                context: self.context,
+            });
```

Comment:
The handling of `queryId` here seems not quite right. First, AFAICT these are supposed to be tracked per-context, rather than globally. And second, Tracy implementations appear to be wrapping this number modulo some sort of a array/pool size. This is important for collecting the timestamps as far as I can tell. My understanding is that these IDs should also be made accessible to the holder of the `GpuSpan` somehow so that they could implement their own timestamp queries.

Reply:
Tracking them per context would be fine - I'll just move to a local atomic.

Why would the implementer need to know the IDs at all? Once they make a gpu span, they create a timestamp query for the beginning of the span, then when they end it they register a timestamp query for the end. Once they both resolve, they call this function to upload the values for those timestamps.